devito/function.py

```diff
@@ -1074,13 +1104,27 @@ def _dist_mask(self):
         return ret
 
     @property
-    def _dist_subfunc_mask(self):
+    def _dist_subfunc_scatter_mask(self):
         """
-        Return a mask, suitable to index into any of self's SubFunctions,
-        usable to reorder a SubFunction values so that adjacent values
-        logically belong to the same MPI rank.
+        This method is analogous to :meth:`_dist_scatter_mask`, although
+        the mask is now suitable to index into self's SubFunctions, rather
+        than into ``self.data``.
         """
-        return self._dist_mask[self._sparse_position]
+        return self._dist_scatter_mask[self._sparse_position]
+
+    @property
+    def _dist_gather_mask(self):
+        """
+        Return a mask to index into the ``data`` received upon returning
+        from ``self._dist_alltoall``. This mask creates a new data array
+        in which duplicate sparse data values have been discarded. The
+        resulting data array can thus be used to populate ``self.data``.
+        """
+        ret = list(self._dist_scatter_mask)
```

Comment:
the received part is not what sent, it should be the complement of dist so 
`mask = np.array(flatten(dmap[i] for i in sorted(dmap) if i != self.rank), dtype=int)`